file.go

```diff
@@ -174,13 +172,21 @@ func (f *win32File) asyncIo(c *ioOperation, deadline time.Time, bytes uint32, er
 		if err == syscall.ERROR_OPERATION_ABORTED {
 			if f.closing {
 				err = ErrFileClosed
-			} else if timedout {
-				err = ErrTimeout
 			}
 		}
-		f.wg.Done()
-		return int(r.bytes), err
+	case <-timeout:
+		cancelIoEx(f.handle, &c.o)
+
+		<-c.ch
+		// runtime.KeepAlive is needed, as c is passed via native
+		// code to ioCompletionProcessor, c must remain alive
+		// until the channel read is complete.
+		runtime.KeepAlive(c)
+
+		err = ErrTimeout
```

Comment:
>err = ErrTimeout [](start = 2, length = 16)

you still need to look at r.err here -- it's possible that the IO completed successfully or with a different error even though you called CancelIoEx